### ISSUE TITLE
Using ebpf-for-windows in production

### ISSUE THREAD

**Issue by zdandoh** [microsoft/ebpf-for-windows#274]
I've been tinkering with this project over the last few days, and think the current functionality would be an awesome fit for a near-term project in my organization. This project is clearly still under heavy development, but it's mentioned in the Getting Started guide that there will be production signed releases. I was wondering if there are any plans for a first release or if you could provide any indication of how production-ready the project is. Thanks!

**Comment by Alan-Jowett** [comment]
Thank you for the feedback. Would it be possible to share out what your scenario is? What hook types are you hoping to use and what helper functions? This information will definitely help ensure guide the project in the right direction.

**Comment by zdandoh** [comment]
So far I've been using the xdp hook to do some packet inspection/dropping. I'd like to also use the bpf_map_* helpers to forward specific packets to a userspace process for further analysis/logging, which then helps make drop decisions for subsequent packets. The feature requested here:
https://github.com/microsoft/ebpf-for-windows/issues/214
Would also be highly valuable.

**Comment by dthaler** [comment]
Production signing is blocked on security hardening as tracked by various other issues.

**Comment by poornagmsft** [comment]
Filed issue 278 to track the helper for copying packet contents into a map.

**Comment by dthaler** [comment]
Per meeting discussion, will create a "security" project that includes any issues that are considered blockers for production signing of the ebpf drivers.
@dthaler to create the project :)

**Comment by dahavey** [comment]
https://github.com/orgs/microsoft/projects/212

**Comment by nicpenning** [comment]
👋🏻 Where did this land? I am curious to how far the signed binaries are and when eBPF for Windows will be production ready (if ever).

**Comment by CedricBouton** [comment]
I am also interested in this progression !

**Comment by olljanat** [comment]
One simple way to handle security concerns would be to release version where unsafe features are disabled/removed and which would allow only whitelisted eBPF programs.

So question for those who are looking for production signed version of this: Have you found production ready open source eBPF programs which already works in Windows?

I’m interested about #4585 but just getting started here so no idea yet how much work there is to get it working in Windows.

EDIT: There seems to be https://opensource.microsoft.com/blog/2022/10/25/towards-debuggability-and-secure-deployments-of-ebpf-programs-on-windows/ talking about topic.

For those who are not familiar with topic, the challenge here is to avoid Bring Your Own Vulnerable Driver (BYOVD) type of attacks. There is a lot of articles about them.

**Comment by nicpenning** [comment]
No, as I am interested from the Elastic eBPF side: https://github.com/elastic/ebpf/issues/188

**Comment by CedricBouton** [comment]
In my case, I want to inspect (or clone?) packets with a specific etherType and retrieve them using XDP sockets. The goal is to implement other protocols over Ethernet.

Unless I am mistaken, there is no mechanism available to do so yet, I didn't found this kind of eBPF program on windows. It seem to be possible on linux with the XDP_REDIRECT way.

**Comment by CedricBouton** [comment]
Do we have any news on this?

Is it possible for us to take care of the signing ourselves?

**Comment by olljanat** [comment]
> Is it possible for us to take care of the signing ourselves?

I have ability to publish Microsoft signed Windows kernel mode drivers at least until June (my company EV code signing certificate expires then and I’m not sure if will renew it because it is quite expensive). See [Container Cpu Affinity](https://github.com/olljanat/windows-containers-cpu-affinity-driver) which I published earlier this week and [Blue Screen Once](https://github.com/olljanat/BlueScreenOnce) which explains more about who I got that ability.

But this goes back to my earlier question: Is there any open sourced eBPF based tools which actually works already in Windows? There is no point to sign non-working beta code.

For XDP you need to use [XDP For Windows](https://github.com/microsoft/xdp-for-windows/). See https://github.com/microsoft/xdp-for-windows/issues/469#issuecomment-1986092227 about how to enable eBPF mode.

EDIT: It seems that next XDP For Windows release will enable eBPF support by default as result of https://github.com/microsoft/xdp-for-windows/pull/961
